locust_plugins/listeners/cloudwatch.py

```diff
@@ -0,0 +1,247 @@
+        if fail_count_rec := self.get_cw_metrics_failure_count_record():
+            cw_metrics_for_request.append(fail_count_rec)
+        return cw_metrics_for_request
+
+    def get_cw_metrics_failure_count_record(self):
+        """
+        Creates a Cloudwatch metric record for FailureCount if the request failed.
+        """
+        dimensions = self.get_metric_dimensions()
+        result = {}
+
+        if self.status == RequestResult.STATUS_FAILURE:
+            result = {
+                "MetricName": "FailureCount",
+                "Dimensions": dimensions,
+                "Timestamp": self.timestamp,
+                "Value": 1,
+                "Unit": "Count",
+            }
+
+        return result
+
+    def get_cw_metrics_response_size_record(self):
+        """
+        Creates a Cloudwatch metric record for Response size if the request succeeded.
+        """
+        dimensions = self.get_metric_dimensions()
+
+        result = {}
+
+        if self.status == RequestResult.STATUS_SUCCESS:
+            result = {
+                "MetricName": "ResponseSize_Bytes",
+                "Dimensions": dimensions,
+                "Timestamp": self.timestamp,
+                "Value": self.response_length,
+                "Unit": "Bytes",
+            }
+
+        return result
+
+    def get_cw_metrics_response_time_record(self):
+        """
+        Creates a Cloudwatch metric record for response time.
+        """
+        dimensions = self.get_metric_dimensions()
+        result = {
+            "MetricName": "ResponseTime_ms",
+            "Dimensions": dimensions,
+            "Timestamp": self.timestamp,
+            "Value": self.response_time,
+            "Unit": "Milliseconds",
+        }
+        return result
+
+    def get_cw_metrics_count_record(self):
+        """
+        Creates a Cloudwatch metric record for request count.
+        """
+        dimensions = self.get_metric_dimensions()
+        result = {
+            "MetricName": "RequestCount",
+            "Dimensions": dimensions,
+            "Timestamp": self.timestamp,
+            "Value": 1,
+            "Unit": "Count",
+        }
+        return result
+
+    def get_metric_dimensions(self):
+        """
+        Provides a collection of generic dimensions which are attached to all cloud watch metrics.
+        Helps in identifying the individual type of request and hence different metrics could be plotted
+        for them.
+        """
+        result = [
+            {"Name": "Request", "Value": self.name},
+            {"Name": "Method", "Value": self.request_type},
+            {"Name": "Host", "Value": self.host},
+        ]
+        return result
+
+    def api_id(self):
+        """
+        A way to uniquely identify the request based on request_type (GET, POST etc.) and
+        the name of the request as given in the locust task.
+        """
+        return f"{self.request_type}-{self.name}"
+
+
+class ServiceContext:
```

Comment:
Does this class really add much value Couldnt all of these things be made properties of the CloudWatchAdapter instead?

Reply:
I just thought that service context for the client is generic and gives a clean namespace for them to differentiate the rest of the stuff. Please let me know if that sound ok to you.

Reply:
I think it is better represented as properties on the CloudWatchAdapter. If this were part of a massive enterprise system it could make sense to have it as an object, but I think it just adds code complexity :)

Reply:
Hmmm... Different ways of thinking I guess. But I will change it. You are the boss on this one. :)

Reply:
Done with the change 👍